core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -86,6 +72,84 @@
         public static final String FETCH_SHARD_STORE = "fetch_shard_store";
     }
 
+    public enum ThreadPoolType {
+        CACHED("cached"),
+        DIRECT("direct"),
+        FIXED("fixed"),
+        SCALING("scaling");
+
+        private final String type;
+
+        public String getType() {
+            return type;
+        }
+
+        ThreadPoolType(String type) {
+            Objects.requireNonNull(type);
+            this.type = type;
+        }
+    }
+
+    public static Map<String, ThreadPoolType> THREAD_POOL_TYPES;
+
+    static {
+        HashMap<String, ThreadPoolType> map = new HashMap<>();
+        map.put(Names.SAME, ThreadPoolType.DIRECT);
+        map.put(Names.GENERIC, ThreadPoolType.CACHED);
+        map.put(Names.LISTENER, ThreadPoolType.FIXED);
+        map.put(Names.GET, ThreadPoolType.FIXED);
+        map.put(Names.INDEX, ThreadPoolType.FIXED);
+        map.put(Names.BULK, ThreadPoolType.FIXED);
+        map.put(Names.SEARCH, ThreadPoolType.FIXED);
+        map.put(Names.SUGGEST, ThreadPoolType.FIXED);
+        map.put(Names.PERCOLATE, ThreadPoolType.FIXED);
+        map.put(Names.MANAGEMENT, ThreadPoolType.SCALING);
+        map.put(Names.FLUSH, ThreadPoolType.SCALING);
+        map.put(Names.REFRESH, ThreadPoolType.SCALING);
+        map.put(Names.WARMER, ThreadPoolType.SCALING);
+        map.put(Names.SNAPSHOT, ThreadPoolType.SCALING);
+        map.put(Names.FORCE_MERGE, ThreadPoolType.FIXED);
+        map.put(Names.FETCH_SHARD_STARTED, ThreadPoolType.SCALING);
+        map.put(Names.FETCH_SHARD_STORE, ThreadPoolType.SCALING);
+        THREAD_POOL_TYPES = Collections.unmodifiableMap(map);
+    }
+    
+    private static void add(Map<String, Settings> executorSettings, ExecutorSettingsBuilder builder) {
+        Settings settings = builder.build();
+        String name = settings.get("name");
+        executorSettings.put(name, settings);
+    }
+    
+    private static class ExecutorSettingsBuilder {
+        Map<String, String> settings = new HashMap<>();
+
+        public ExecutorSettingsBuilder(String name) {
+            settings.put("name", name);
+            settings.put("type", THREAD_POOL_TYPES.get(name).getType());
+        }
+
+        public ExecutorSettingsBuilder size(int availableProcessors) {
+            return add("size", Integer.toString(availableProcessors));
+        }
+        
+        public ExecutorSettingsBuilder queueSize(int queueSize) {
+            return add("queue_size", Integer.toString(queueSize));
```

Comment:
As an outside (separate issue), I think we should bound this at a reasonable level to avoid people setting it waaaay too high, perhaps we can open an issue to limit it here to < 100,000?